src/track/beatmap.cpp

```diff
@@ -752,3 +768,84 @@ double BeatMap::calculateBpm(const Beat& startBeat, const Beat& stopBeat) const
 
     return BeatUtils::calculateBpm(beatvect, m_iSampleRate, 0, 9999);
 }
+
+mixxx::TimeSignature BeatMap::getSignature(double dSample) const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid())
+        return mixxx::kNullTimeSignature;
+
+    auto result = mixxx::kDefaultTimeSignature;
+
+    // Special case, when looking for initial TimeSignature
+    if(dSample == 0) {
+        auto beat = m_beats.cbegin();
+        if (beat->has_signature()) {
+            result.setBeats(beat->signature().beats());
+            result.setNoteValue(beat->signature().note_value());
+        }
+    } else {
+        // Scans the list of beats to find the last time signature change before the sample
+        for( auto beat = m_beats.begin() ; beat != m_beats.end() && beat->frame_position() < dSample ; beat++) {
+           if(beat->has_signature()) {
+               result.setBeats(beat->signature().beats());
+               result.setNoteValue(beat->signature().note_value());
+           }
+        }
+    }
+    return result;
+}
+
+void BeatMap::setSignature(mixxx::TimeSignature sig, double dSample) {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid())
+        return;
+
+    // Moves to the beat before the sample
+    BeatList::iterator beat = m_beats.begin();
+    for(; beat != m_beats.end() && beat->frame_position() < dSample ; ++beat)
+        ;
+
+    // If at the end, change nothing
+    if ( beat == m_beats.end()) {
+        return;
+    }
+
+    // Adjust position if not at the first beat
+    if (beat != m_beats.begin()) {
+            beat--;
+    }
+
+    // Sets the TimeSignature value
+    beat->mutable_signature()->set_beats(sig.getBeats());
+    beat->mutable_signature()->set_note_value(sig.getNoteValue());
+    locker.unlock();
+    emit(updated());
+}
+
+void BeatMap::setBar(double dSample) {
```

Comment:
What is the rationale behind this function?
Why does it break in the next phrase? Why does it call makeBeat?

Reply:
This function sets the next beat as a bar beat (or down beat) and then navigates through the rest of the beats list, until the end or until the next phrase beat, setting all the beats as normal beats (makeBeat) but the ones that are a multiple of a bar that are marked as bar beats (makeBar).
I stop at the next phrase, if there's any, because I consider that a phrase always resets the bar count.
Is this explanation clear enough?

Reply:
Thanks for your reply, Javier but I am still confused...

The explanation of the code was clear, but that's not what I was confused with, it's actually with the decisions behind it. Probably, it was my comment that wasn't very clear I guess. I will try again.

Is it really necessary to set all the other beats as normal beats?
Beforehand, you are making all the beats in the beat list of type beat by default right?
Why set then again as type beat with makeBeat? Or have I missed something? 
And from my understanding, even the downbeats, and phrase beats are still a regular beat none the less.. My point being, all the beats in the beat list should be in the first place a normal beat, so there shouldn't be any need to set this.

And I still don't get why to stop in the next phrase.
What do you mean by "resets the bar count"? By bar count, you meant the number of beats in the bar? Then ok, I understand why to stop. However still, why not stop when the time signature actually changes, instead of assuming that it will change with all phrases?

Otherwise, if you meant something else by bar count like actual bar counting, I don't get it.

I just want to make sure to clearly understand your thoughts on this and how you are approaching the problem, so we can agree on an interface that I could build upon.

Reply:
Ok. So there are two points to clarify.

1. Why I reset all the bar beats after the newly set up bar beat.
2. Why I stop in that reset if I find a phrase beat.

The short answer is that I made it that way due to the way a DJ, at least me and the few I know, will prepare a track after importing it into the software. Even if we have an analyzer that is able to identify bar beats and phrase beats, I will always manually beatgrid the track, I will not accept the risk that involves to blindly trust the analyzer result.

Point 1:
Imagine a user has added a new track (song) to his library, imagine this track starts with 4/4 tempo and it has a short bridge bar with only  three in the middle. After analyzing it, the user  wants to beatgrid it.

The user sets a bar beat at the track very beginning. Immediately the algorithm sets a bar beat at every four beats. If the beat bars are nor set automatically, the user will have to set manually every single bar beat in the track, a tedious work.

It becomes like 1b-2-3-4-5b-6-7-8-b9-10-11-12-b13-14-15-16-b17-18-19-20 and so on. Bar beats are indicated by a b.

When the user listens to the track for beatgridding it, setting cue points, etc, he identifies the short bridge bar at beat nine. So he sets a new bar beat at the third beat of that bridge. If the algorithm only adds this bar beat and the ones after that point, the track will be messed and will look like 1b-2-3-4-5b-6-7-8-b9-10-11-B12-b13-14-15-B16-b17-18-19-B20. The new bar beats are indicated by B.

This is why I opted for "resetting" the bars, to get 1b-2-3-4-5b-6-7-8-b9-10-11-B12-13-14-15-B16-17-18-19-B20.

Point 2:
In general, I first look at the phrases in the track. I play almost only electronic music, so I will look for breakdowns, drops, and climax phrases. I assume that a phrase always starts in a bar. So the phrase beats identified are already good quality bar beats. Usually short bars happen at the end of a phrase.

So I don't want to destroy this good quality bar beats already identified and this is why I stop the algorithm when I find one of this phrase beats.

In the previous example, there is high probability that the beat number 12 is also a phrase beat.

Is this explanation making more clear?

I can agree that all this makes sense in the context of a manual workflow for something automated like an analyzer, may be is not so convenient. But I consider that the software must support the manual workflows. 

What can be discussed is if the workflow is sound or not. If we can't agree on it may be is the moment to make a more open question in Zulip to see what others think about.

Sorry for the length of this post, but I have problems to better express :)

Reply:
No worries about the length of the post, if anything, thanks for putting the effort to make a clear explanation. Everything makes a whole lot more sense to me now and I understand your decisions, from the perspective of the user and manual intervention they are very reasonable. I think the workflow is sound. 
It's now clear for me, that we are going to need 2 different interfaces, one for the detection algorithm and another one for manual intervention. I think the iterator interface you made is enough for the algorithm though so that's already done. 
But then we have a new and very tricky problem for the manual intervention implementation, that is to do with the detected bars and/or phrases after there is user intervention. 
Your decision to stop in the next phrase is actually probably something smart to do in the case of the user adjusting a bar position. 
But there are a lot of possible cases and things that the user might expect, a discussion in Zulip is probably needed to discuss this. I will start to think about it and write something there when I have some clear thoughts on this issue.

Reply:
> But then we have a new and very tricky problem for the manual intervention implementation, that is to do with the detected bars and/or phrases after there is user intervention.

Yes, it's a hard problem. We'll have to discuss how to handle this on Zulip.

Reply:
Yes, I agree that having two different ways of setting bar an phrase beat makes sense. One for humans taking into account the user workflow and a different one for algorithms, like the analyzer or the first way mentioned above.
As right now we have no analyzer to identify bar and phrase, I will concentrate in ensuring a good workflow for the user. At some point, when the analyzer is ready to be used we will have to look at the implications it has for the user workflow.